File: Password.java
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Scanner;

/**
 * Password class contains password-related operations such as encrypting,
 * decrypting, validating, and resetting the password file.
 * @author Debra-Lee Speight
 */
public class Password{
	private final Errors errors;
	private final Scanner in;
	private Scanner fileScan;
	private File file;

	/**
	 * Constructor instantiates class members.
	 */
	Password(){
		errors = new Errors();
		in = new Scanner(System.in);
		fileScan = null;
		file = null;
	}

	/**
	 * Allows user to reset password upon validation of current password.
	 */
	void resetPW(){
		String newPW = "newPW", confirmNewPW = "confirmPW";
		boolean oldPWvalidated = validatePW();
		while(oldPWvalidated && !newPW.equals(confirmNewPW)){
			System.out.print("\n\tEnter new password:  ");
			newPW = in.nextLine();

			System.out.print("\tConfirm new password:  ");
			confirmNewPW = in.nextLine();

			if(newPW.equals(confirmNewPW)){
				try{file.createNewFile();}
				catch(IOException e){errors.errorMessage("create", "", "");}

				try{
					FileWriter fileWriter = new FileWriter(file);
					fileWriter.write(encryptPW(newPW));
					fileWriter.close();
				}
				catch(IOException e){errors.errorMessage("filewriter", "", "");}
			}
			else System.out.println("\n\t(!) ERROR: The passwords do not match.");
		}
		System.out.println("\n\tPassword changed successfully.");
	}

	/**
	 * Validates the password by comparing the decrypted password to the user's
	 * password input; user gets 3 attempts to enter the correct password.
	 * @return true if password entered is correct; false otherwise
	 */
	boolean validatePW(){
		String currentPW = getCurrentPW();
		for(int i = 0; i < 3; i++){
			System.out.print("\n\tENTER CURRENT PASSWORD:  ");
			String pwEntered = in.nextLine();

			if(pwEntered.equals(currentPW)){
				System.out.println("\n\tACCESS GRANTED!");
				return true;
			}
			else System.out.println("\n\tINVALID PASSWORD");
		}
		System.out.println("\n\tACCESS DENIED!");
		return false;
	}

	/**
	 * Creates the password file, encrypts and writes the default password
	 * to the password file, and displays the default password to the user.
	 */
	private void setDefaultPW(){
		try{
			FileWriter fileWriter = new FileWriter(file);
			fileWriter.write(encryptPW(" "));
			fileWriter.close();
		}
		catch(IOException e){errors.errorMessage("filewriter", "", "");}
		errors.errorMessage("defaultPW", "", "");
	}

	/**
	 * Opens the password file, then decrypts and returns the password.
	 * @return currentPW decrypted
	 */
	private String getCurrentPW(){
		file = new File("PW");
		if(!file.exists()){
			try{file.createNewFile();}
			catch(IOException e){errors.errorMessage("create", "", "");}
			setDefaultPW();
		}
		try{fileScan = new Scanner(file);}
		catch(IOException e){errors.errorMessage("scanner", "", "");}

		String currentPW = fileScan.nextLine();
		fileScan.close();
		return decryptPW(currentPW);
	}

	/**
	 * Encrypts the password in the password text file for security.
	 * @param pw
	 * @return encryptedPW
	 */
	private String encryptPW(String pw){
		int length = pw.length();
    	String encryptedPW = "";
		int i = 0;
		while(i != length){
			Character character = pw.charAt(i);
			int ascii = character + 12;
			encryptedPW += (char)ascii;
			i++;
		}
		return encryptedPW;
	}

	/**
	 * Decrypts the password in the password text file for validation.
	 * @param encryptedPW
	 * @return decryptedPW
	 */
	private String decryptPW(String encryptedPW){
		int length = encryptedPW.length();
    	String decryptedPW = "";
		int i = 0;
		while(i != length){
			Character character = encryptedPW.charAt(i);
			int ascii = character - 12;
			decryptedPW += (char)ascii;
			i++;
		}
		return decryptedPW;
	}
}

File: Password.c
#include "Header.h"

/********************************************************************************
 *	Password.c file contains password-related operations such as encrypting,	*
 *	decrypting, validating, and resetting the password file.					*
 *																				*
 *	@author Debra-Lee Speight													*
 ********************************************************************************/

/**
 * Allows user to reset password upon validation of current password;
 * continues asking for new password and confirmation of new password
 * until the two are a match.
 */
void resetPW(){
	char newPW[50] = "newPW";
	char confirmNewPW[50] = "confirmNewPW";
	bool oldPWvalidated = validatePW(), newPWvalidated = false;

	FILE *PWfile = fopen("PW", "w+");
	if(PWfile == NULL) fileCreationError("PASSWORD FILE");

	while(oldPWvalidated && !newPWvalidated){
		printf("\n\tEnter new password:  ");
		scanf("%s", newPW);

		printf("\tConfirm new password:  ");
		scanf("%s", confirmNewPW);

		if(strcmp(newPW, confirmNewPW) == 0){
			newPWvalidated = true;
			fputs(encryptPW(newPW), PWfile);
			fclose(PWfile);
			printf("\n\tPassword changed successfully.\n");
		}
		else printf("\n\t(!) ERROR: The passwords do not match.\n");
	}
}

/**
 * Validates the password by comparing the decrypted password to the user's
 * password input; user gets 3 attempts to enter the correct password.
 * @return true if password entered is correct; false otherwise
 */
bool validatePW(){
	char pwEntered[50];
	char *currentPW = (char*)calloc(50, sizeof(char));
	currentPW = getCurrentPW();
	decryptPW(currentPW);

	for(int i = 0; i < 3; i++){
		printf("\n\tENTER CURRENT PASSWORD:  ");
		scanf("%s", pwEntered);
		if(strcmp(pwEntered, currentPW) == 0){
			printf("\n\tACCESS GRANTED!\n");
			encryptPW(currentPW);
			return true;
		}
		else printf("\n\tINVALID PASSWORD\n");
	}
	printf("\n\tACCESS DENIED!\n");
	encryptPW(currentPW);
	return false;
}

/**
 * Opens the password file, and returns the password encrypted;
 * if no password file exists, a default password is set before returning it
 * (see setDefaultPW()).
 * @return currentPW (encrypted)
 */
const char * getCurrentPW(){
	FILE *fileRead = fopen("PW", "r");
	if(fileRead == NULL){
		fclose(fileRead);
		setDefaultPW();
		fileRead = fopen("PW", "r");
		if(fileRead == NULL) fileReadingError("PW");
	}
	static char currentPW[50];
	fgets(currentPW, 50, fileRead);
	if(currentPW[0] == '.' + 9) defaultPWalert();
	fclose(fileRead);
	return currentPW;
}

/**
 * Creates a new password file, encrypts the default password, and writes the
 * encrypted default password to the password file.
 */
void setDefaultPW(){
	FILE *fileWrite = fopen("PW", "w+");
	if(fileWrite == NULL) fileCreationError("PASSWORD FILE");
	char defPW[2] = ".";
	defPW[0] += 9;
	fputs(defPW, fileWrite);
	fclose(fileWrite);
}

/**
 * Encrypts the password in the password text file for security.
 * @param password (decrypted)
 * @return password (encrypted)
 */
const char * encryptPW(char *password){
	for(int i = 0; i < strlen(password); i++)
		if(password[i] != '\n' && password[i] != 32)
			password[i] += 9;
	return password;
}

/**
 * Decrypts the password in the password text file for validation.
 * @param password (encrypted)
 * @return password (decrypted)
 */
const char * decryptPW(char *password){
	for(int i = 0; i < strlen(password); i++)
		if(password[i] != '\n' && password[i] != 32)
			password[i] -= 9;
	return password;
}